### PR TITLE
cython-based C++-/Python interface

### LINKED ISSUE #217: Python API
To allow calls from C++ to Python and vice versa, we need an awesome interface.

@mic-e started working on a nice example here: https://github.com/mic-e/cythondemo

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8374506-python-api?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #229: Make the code compliance Makefile targets selective
`make checkfull` takes roughly 10 seconds to run here.  If there was a way to only check files that were modified, it would be more practical to run code compliance checks as a commit hook.  That would lead to less frustration with pull requests failing to build because of things like copyright headers not being up to date.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8510466-make-the-code-compliance-makefile-targets-selective?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mic-e**
- openage is now a shared library
 - moved `py/openage` -> `openage`
 - moved `py/openage/codecompliance` -> `buildsystem/codecompliance`
 - moved `cpp/` -> `libopenage/`
 - moved `doc/implementation/` -> `doc/code`
 - integrated Cython into buildsystem
 - C++ -> Py function calling
  - C++ API is specified in PXD annotations in `.h` files
  - cpp/pyinterface/pyfunc.h
   - Func<returntype, arg0type, arg1type, ...>
   - Callable objects
   - Designed to be filled with pointers to cdef functions
   - __don't use raw function pointers in the interface__
  - openage/cppinterface/setup.py
   - setup() Fills the PyFunc pointers
   - must be called before using libopenage functionality
 - Exception translation
  - cpp/pyinterface/exctranslate.h
   - translate_exc_cpp_to_py
   - translate_exc_py_to_cpp
  - py/cppinterface/exctranslate.py
   - describe_exception
   - raise_cpp_error
  - C++ functions that are declared 'except +' if they may throw
  - Python functions that are declared 'except *' if they may raise
  - GIL-safe
  - everything is translated back and forth
   - metadata
   - message
   - backtrace
 - Documentation

C++:

 - Overhauled exception code
  - now lives in `error/error.h`
  - exceptions automatically store their cause exception
  - exceptions can collect backtraces
  - see demo in `cpp/error/demo.cpp`
  - Hierarchies may now be used
  - `doc/code/exceptions`
 - Moved util/error.h to error/error
 - Added error/handlers.cpp, which provides a SIGSEGV and terminate handler
 - refactored LZXD code
  - removed many C-isms and micro-optimizations
  - simplified codeflow, split methods into smaller ones
  - file split into multiple smaller ones (`cpp/util/compress`)
  - `decompress` replaced by `decompress_frame`, which decompresses just
    32768 bytes at a time instead of everything at once.
  - exposed to Python using the PyInterface
 - Simplified C++ logging
  - ExternalStringStream: like stringstream, but appends to a custom buffer
 - Modernized C++ testing
  - now uses TestError (e.g. via the `TESTFAIL` macro) instead of return values.
 - C++ exceptions can collect backtraces
 - `SIGSEGV` and `std::terminate` handlers
 - Enum class that allows adding new members at runtime, and associated data.
  - used by `log::lvl`
 - `util/init.h`: `OnInit` and `OnDeInit`: to call functions on dynamic initialization
 - fixed a bug with Timer

Python:

 - Bumped Python version requirement to 3.4
 - Massive clean-up
  - Unified all utility modules into the openage.util package
  - Removed some code that got obsoleted in the process
  - Made stuff more pylint-compliant
  - Replaced home-brewn Enum and NamedObject implementations by enum.Enum
 - Implemented some utilities
  - openage/util/fslike.py
   - FileSystemLikeObject abstract base class
   - Directory FileSystemLikeObject
  - openage/util/filelike.py
   - FileLikeObject abstract base class
   - StreamSeekBuffer FileLikeObject
   - StreamFragment FileLikeObject
   - FIFO FileLikeObject
  - FSLikeObject used by convert and codegen
 - Overhauled testing system
  - openage.testing.testing.assert_result
  - openage.testing.testing.assert_raises
  - openage.testing.testing.TestError
  - now done without CTest
  - C++ and Python tests and demos all called via Python
 - Integrated cabextract/cab.py with the rest of the project
  - now a FileSystemLikeObject
  - takes a FileLikeObject as an argument
  - added tests
 - PEFile has been refactored
  - resource strings are now resolved
  - the code uses NamedStructs
 - The Python logger now forwards to the C++ logger if the interface has been
   initialized
  - Central management of current loglevel
   - `--verbose`, `--quiet`
 - Program has one single entry point: `openage.__main__.main()`
  - run with `python3 -m openage`
  - alternatively: `./run`
  - subcommands: `game`, `test`, `codegen`
 - Media file conversion is now done when launching the game.

Buildsystem:

 - added valgrind-python.supp, a valgrind suppression file.
 - added basic whitespace compliance checking for textfiles
 - added pylint code compliance checking
  - Many files are not yet pylint-compliant; those have a header
    '# TODO pylint: disable=C,R,whatever categories'
 - added selective code compliance checking (closes #229)
 - re-worked Makefile targets

closes #217 

See the various doc files, including [pyinterface](https://github.com/mic-e/openage/blob/pyinterface/doc/code/pyinterface.md), [testing](https://github.com/mic-e/openage/blob/pyinterface/doc/code/testing.md)

Some things are still remaining to do, but they in no way prevent merging:

 - write interface for `FSLikeObject`
  - use that interface for all asset handling (in C++)
  - add `inotify` support to `FSLikeObject`
 - support for other sources for `openage.convert` (setup disk/.iso #82)

Note: If you wish to start work that derives from this before it's been merged, which you should, if you're about to start any new branch, base it on [this commit](https://github.com/mic-e/openage/commit/483c482649fa026cee8fe4024237f34f23255ab3). That commit is guaranteed to end up as-is in `openage/master`.

### CODE REVIEW

**Comment by franciscod** [comment]
nice! :tada:

**Comment by mic-e** [comment]
Obviously related: #217

**Comment by mic-e** [comment]
And thus, it has been completed.

**Comment by mic-e** [comment]
I need some OSX user to test this, please.

**Comment by JonasBorchelt** [comment]
i will test it later

**Comment by TheJJ** [comment]
`+16,667 −3,710`, minor changes, :yum:...

File: buildsystem/codecompliance/authors.py [hunk 1]
```diff
@@ -14,9 +15,9 @@ def get_author_emails_copying_md():
 
     |     name     |    nick    |    email    |
     """
-    with open("copying.md") as f:
-        for line in f:
-            match = re.match("^.*\|[^|]*\|[^|]*\|([^|]*)\|.*$", line)
+    with open("copying.md") as fobj:
+        for line in fobj:
+            match = re.match("^.*\\|[^|]*\\|[^|]*\\|([^|]*)\\|.*$", line)
```
Comment (TheJJ) [inline buildsystem/codecompliance/authors.py]: just write `r"..."` to avoid the backslash escaping.

File: buildsystem/codecompliance/authors.py [hunk 2]
```diff
@@ -45,7 +46,7 @@ def get_author_emails_git_shortlog(exts=('.cpp', '.h', '.py', '.cmake')):
     output = Popen(invocation, stdout=PIPE).communicate()[0]
 
     for line in output.decode('utf-8', errors='replace').split('\n'):
-        match = re.match("^ +[0-9]+\t[^<]*\<(.*)\>$", line)
+        match = re.match("^ +[0-9]+\t[^<]*\\<(.*)\\>$", line)
```
Comment (TheJJ) [inline buildsystem/codecompliance/authors.py]: `r"..."`

File: buildsystem/codecompliance/pystyle.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2014-2015 the openage authors. See copying.md for legal info.
+
+"""
+Checks PEP8 compliance, with some exceptions.
+"""
+
+from pep8 import StyleGuide
+
+# these errors will be ignored by pep8
+IGNORE_ERRORS = "E251,E221,E501,E241"
```
Comment (TheJJ) [inline buildsystem/codecompliance/pystyle.py]: we should probably add a comment what these errors are about.

File: doc/building.md [hunk 4]
```diff
@@ -34,12 +34,13 @@ Dependencies are needed for:
 
 * C = compiling
 * R = running
-* M = media convert script
+* M = media conversion (when running for the first time)
```
Comment (TheJJ) [inline doc/building.md]: i'd remove the "first time" thing as we'll need reconverts.

File: buildsystem/codecompliance/authors.py (hunk 2 above)
Comment (mic-e) [inline buildsystem/codecompliance/authors.py]: Then how do I escape `\t`?

File: buildsystem/python.cmake [hunk 5]
```diff
@@ -7,176 +7,160 @@
 # for an example usage, see py/openage/convert/CMakeLists.txt
 
 function(python_init)
-	find_package(Python 3.3 REQUIRED)
+	find_package(Python 3.4 REQUIRED)
```
Comment (TheJJ) [inline buildsystem/python.cmake]: i think this requires a newer (3.0.0) cmake version (in `building.md` and in the `/cmakelists.txt`)

determined by: `git tag --contains=ab6201ab`
Kitware/Cmake@ab6201ab647c13fee44c44822b8502e9848faaaf

File: buildsystem/codecompliance/authors.py (hunk 2 above)
Comment (TheJJ) [inline buildsystem/codecompliance/authors.py]: uh, yeah, you're right..

File: buildsystem/simple [hunk 6]
```diff
@@ -17,16 +17,14 @@ PYVERDOTLESS=34m
...
-# these files would be filled with actual test lists by testing.cmake
-touch assets/tests_py assets/tests_cpp
 # this file would be generated by cmake
 echo "global_asset_dir = '/usr/share/openage'" > py/openage/config.py
 echo "version = '`cat openage_version`'" >> py/openage/config.py
 
-# generate C code from convert script
-PYTHONPATH="$PYTHONPATH:py" python3 -m openage.codegen --target-cache=/tmp/codegentargets --depend-cache=/tmp/codegendepends --cpp-src-dir=cpp --write-to-sourcedir
+# generate C++ code from convert script
+PYTHONPATH="$PYTHONPATH:py" python3 -m openage codegen --target-cache=/tmp/codegentargets --depend-cache=/tmp/codegendepends --cpp-src-dir=cpp --write-to-sourcedir
```
Comment (TheJJ) [inline buildsystem/simple]: i think it's `projectdir` instead of `cpp-src-dir` now right?

File: buildsystem/warn_runtime_depends.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    "PIL.ImageDraw",
+    "numpy"
+]
+
+
+def depend_error(text):
+    """
+    prints a nice and colorful warning message.
+    """
+    print("\x1b[1;33mWARNING:\x1b[m unmet runtime dependency: {}".format(text))
```
Comment (TheJJ) [inline buildsystem/warn_runtime_depends.py]: `isatty`?

File: configure [hunk 8]
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env python3
+
+# Copyright 2013-2015 the openage authors. See copying.md for legal info.
+
+# Together with the Makefile, ./configure provides an autotools-like build
+# experience. For more info, see --help and doc/buildsystem.
+
 import sys
 if not sys.version_info >= (3, 3):
     print("openage requires Python 3.3 or higher")
```
Comment (TheJJ) [inline configure]: we require 3.4 now

File: buildsystem/simple (hunk 6 above)
Comment (mic-e) [inline buildsystem/simple]: the entire `simple` file is broken anyways...

Comment (TheJJ) [inline buildsystem/simple]: then kill it, if it doesn't work any more.

File: cpp/CMakeLists.txt [hunk 9]
```diff
@@ -79,6 +80,19 @@ find_package(SDL2Image REQUIRED)
 find_package(Opusfile REQUIRED)
 pkg_search_module(EPOXY REQUIRED epoxy)
 
+target_link_libraries(${PROJECT_NAME} dl rt)
```
Comment (TheJJ) [inline cpp/CMakeLists.txt]: please add `dl` and `rt` to the main link library section below.

Comment (mic-e) [inline cpp/CMakeLists.txt]: which one?

Comment (TheJJ) [inline cpp/CMakeLists.txt]: line `145` in this file, in
 ```cmake
target_link_libraries(${PROJECT_NAME}
 	${FONTCONFIG_LIB}
	...
```

File: cpp/error/backtrace.h [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	 * The most recent call is returned last (alike Python).
+	 *
+	 * @param cb
+	 *    is called for every symbol in the backtrace, starting with the top-most
+	 *    frame.
+	 * @param reversed
+	 *    if true, the most recent call is given last.
+	 */
+	virtual void get_symbols(std::function<void (const backtrace_symbol *)> cb,
+	                         bool reversed = true) const = 0;
```
Comment (TheJJ) [inline cpp/error/backtrace.h]: `reversed=true`, no spaces for default values.

File: cpp/player.h [hunk 11]
```diff
@@ -46,6 +46,6 @@ class Player {
 
 };
 
-}
+} // player.h
```
Comment (TheJJ) [inline cpp/player.h]: `// openage`

File: buildsystem/codecompliance/textfiles.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+        if ' \n' in data or '\t\n' in data:
+            if filename.startswith('openage/') and filename.endswith('.cpp'):
+                # allow this particular error for Cython-generated files.
+                pass
+            elif filename.endswith('.gen.h') or filename.endswith('.gen.cpp'):
+                # TODO all this for now, until someone fixes the codegen.
+                pass
+            else:
+                yield "Trailing whitespace", filename
```
Comment (TheJJ) [inline buildsystem/codecompliance/textfiles.py]: maybe we can also print the linenumber, but only calculate it when the failure was found. at least add a TODO for this.

File: CMakeLists.txt [hunk 13]
```diff
@@ -58,29 +57,39 @@ include("${BUILDSYSTEM_DIR}/doxygen.cmake")
...
 warn_runtime_depends()
+
+# don't print 'Built target ...' messages
+# this will hopefully be merged into CMake 3.3; until then, this will have
+# no effect.
+# http://www.cmake.org/pipermail/cmake-developers/2015-July/025649.html
+set_property(GLOBAL PROPERTY TARGET_MESSAGES OFF)
+
+# make compile command metadata available (e.g. to clang-tools)
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
Comment (TheJJ) [inline CMakeLists.txt]: redundant with `buildsystem/cpp.cmake`

File: buildsystem/cpp.cmake [hunk 14]
```diff
@@ -5,6 +5,7 @@
 #TODO: integrate PGO (profile-guided optimization) build
 
 function(cpp_init)
+	set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
Comment (TheJJ) [inline buildsystem/cpp.cmake]: redundant with the root `CMakeLists.txt`, we probably want to keep it here though i guess.

**Comment by mic-e** [comment]
This is now ready for merging.

**Comment by franciscod** [comment]
um... how do you say it... what about win32? :stuck_out_tongue_winking_eye:

**Pull request merged**